lib/generator_graph.ml

```diff
@@ -0,0 +1,149 @@
+let copy_matrix matrix =
+  Array.init (Array.length matrix) (fun i -> Array.copy matrix.(i))
+
+let check_neighbors_does_not_have_same_color grid height width current_row
+    current_col color =
+  let neighbors =
+    [
+      (current_row - 1, current_col);
+      (current_row + 1, current_col);
+      (current_row, current_col - 1);
+      (current_row, current_col + 1);
+    ]
+  in
+  List.for_all
+    (fun (row, col) ->
+      if row >= 0 && col >= 0 && row < height && col < width then
+        grid.(row).(col) <> color
+      else true)
+    neighbors
+
+let rec generate_colorings grid width height current_row current_col max_colors
+    =
+  if current_row = height then [ copy_matrix grid ]
+  else if current_col = width then
+    generate_colorings grid width height (current_row + 1) 0 max_colors
+  else
+    let rec for_all_color current_color acc =
+      if current_color > max_colors then acc
+      else if
+        check_neighbors_does_not_have_same_color grid height width current_row
+          current_col current_color
+      then
+        let () = grid.(current_row).(current_col) <- current_color in
+        let copy =
+          generate_colorings grid width height current_row (current_col + 1)
+            max_colors
+        in
+        let () = grid.(current_row).(current_col) <- 0 in
+        for_all_color (current_color + 1) (acc @ copy)
+      else for_all_color (current_color + 1) acc
+    in
+    for_all_color 1 []
+
+let generate_all_colorings width height max_colors =
+  let grid = Array.make_matrix width height 0 in
+  generate_colorings grid width height 0 0 max_colors
+
+let check_graphs_differ_by_one_value_on_middle arr1 arr2 =
+  if Array.length arr1 <> Array.length arr2 then false
+  else
+    let rec count_differences i j cmp =
+      if cmp >= 2 then false
+      else if j >= Array.length arr1 then cmp = 1
+      else if i >= Array.length arr1.(j) then count_differences 0 (j + 1) cmp
+      else if arr1.(j).(i) <> arr2.(j).(i) then
+        if i = 1 && j = 1 then count_differences (i + 1) j (cmp + 1) else false
+      else count_differences (i + 1) j cmp
+    in
+    count_differences 0 0 0
+
+let count_number_tile_a_and_b set_tiles set_tilesA set_tilesB =
+  List.fold_left
+    (fun acc e ->
+      if List.exists (fun x -> x = e) set_tilesA then acc + 1
+      else if List.exists (fun x -> x = e) set_tilesB then acc + 1
+      else acc)
+    0 set_tiles
+
+let is_parity_between_tiles_A_and_B graph set_tilesA set_tilesB =
+  let tiles_from_graph =
+    match graph with
+    | [| [| a; b; c |]; [| d; e; f |]; [| g; h; i |] |] ->
```

Comment:
there is toroidal tiles who miss in this : : [[g; h]; [a; b]] for example

Reply:
Not precisely the teacher specified that they were not